File: sandbox-runner/patterns/webhooks/python/starter.py
import asyncio
import time

from temporalio.client import Client

from shared import TASK_QUEUE, WORKFLOW_ID_PREFIX, OrderInput, PaymentPayload
from workflows import OrderWorkflow


async def main() -> None:
    client = await Client.connect("localhost:7233")

    order_id = f"{int(time.time() * 1000)}"
    workflow_id = f"{WORKFLOW_ID_PREFIX}-{order_id}"
    order = OrderInput(order_id=order_id, amount=99.99)

    print(f"Starting order workflow: {workflow_id}")

    # Start the workflow — it will block waiting for the payment_received signal
    handle = await client.start_workflow(
        OrderWorkflow.run,
        order,
        id=workflow_id,
        task_queue=TASK_QUEUE,
    )

    # Simulate the inbound webhook arriving 3 seconds later.
    # In production this would be your HTTP handler calling signal_workflow.
    print("Simulating inbound payment webhook in 3 seconds...")
    await asyncio.sleep(3)

    payment = PaymentPayload(
        payment_id=f"pay-{int(time.time() * 1000)}",
        amount=99.99,
    )

    # This is what your HTTP /webhook handler does upon receiving the POST:
    await client.get_workflow_handle(workflow_id).signal(
        OrderWorkflow.payment_received, payment
    )
    print(f"Webhook signal sent: {payment.payment_id}")

    result = await handle.result()
    print(f"Order completed: {result}")
    print(f"Open the Temporal UI and search for '{workflow_id}' to see the history.")


if __name__ == "__main__":
    asyncio.run(main())

File: sandbox-runner/patterns/webhooks/java/src/main/java/Starter.java
import io.temporal.client.WorkflowClient;
import io.temporal.client.WorkflowOptions;
import io.temporal.client.WorkflowStub;
import io.temporal.serviceclient.WorkflowServiceStubs;

public class Starter {
    public static void main(String[] args) throws Exception {
        WorkflowServiceStubs service = WorkflowServiceStubs.newLocalServiceStubs();
        WorkflowClient client = WorkflowClient.newInstance(service);

        String orderId = String.valueOf(System.currentTimeMillis());
        String workflowId = Shared.WORKFLOW_ID_PREFIX + "-" + orderId;

        OrderWorkflow workflow = client.newWorkflowStub(
                OrderWorkflow.class,
                WorkflowOptions.newBuilder()
                        .setTaskQueue(Shared.TASK_QUEUE)
                        .setWorkflowId(workflowId)
                        .build());

        System.out.println("Starting order workflow: " + workflowId);

        // Start the workflow asynchronously — it will block waiting for the signal
        WorkflowClient.start(workflow::run, new Shared.OrderInput(orderId, 99.99));

        // Simulate the inbound webhook arriving 3 seconds later.
        // In production this would be your HTTP handler calling signalWorkflow.
        System.out.println("Simulating inbound payment webhook in 3 seconds...");
        Thread.sleep(3000);

        Shared.PaymentPayload payment = new Shared.PaymentPayload(
                "pay-" + System.currentTimeMillis(), 99.99);

        // This is what your HTTP /webhook handler does upon receiving the POST:
        workflow.paymentReceived(payment);
        System.out.println("Webhook signal sent: " + payment.paymentId());

        // Wait for the workflow to complete
        String result = WorkflowStub.fromTyped(workflow).getResult(String.class);
        System.out.println("Order completed: " + result);
        System.out.println("Open the Temporal UI and search for '" + workflowId + "' to see the history.");

        System.exit(0);
    }
}
